array.c

```diff
@@ -3310,19 +3310,58 @@ rb_ary_rotate(VALUE ary, long cnt)
 
 /*
  *  call-seq:
- *     ary.rotate!(count=1)   -> ary
+ *    array.rotate! -> self
+ *    array.rotate!(count) -> self
  *
- *  Rotates +self+ in place so that the element at +count+ comes first, and
- *  returns +self+.
+ *  Rotates +self+ in place by moving elements from one end to the other; returns +self+.
  *
- *  If +count+ is negative then it rotates in the opposite direction, starting
- *  from the end of the array where +-1+ is the last element.
+ *  Argument +count+, if given, must be an
+ *  {Integer-convertible object}[doc/implicit_conversion_rdoc.html#label-Integer-Convertible+Objects].
  *
- *     a = [ "a", "b", "c", "d" ]
- *     a.rotate!        #=> ["b", "c", "d", "a"]
- *     a                #=> ["b", "c", "d", "a"]
- *     a.rotate!(2)     #=> ["d", "a", "b", "c"]
- *     a.rotate!(-3)    #=> ["a", "b", "c", "d"]
+ *  ---
+ *
+ *  When no argument given, rotates the first element to the last position:
+ *    a = [:foo, 'bar', 2, 'bar']
+ *    a1 = a.rotate!
+ *    a1 # => ["bar", 2, "bar", :foo]
+ *    a1.equal?(a1) # => true # Retruned self
+ *
+ *  ---
+ *
+ *  When given a non-negative +count+, rotates +count+ elements from the beginning to the end:
+ *    a = [:foo, 'bar', 2]
+ *    a.rotate!(2)
+ *    a # => [2, :foo, "bar"]
+ *
+ *  If +count+ is large, uses <tt>count % ary.size</tt> as the count:
+ *    a = [:foo, 'bar', 2]
+ *    a.rotate!(20)
+ *    a # => [2, :foo, "bar"]
+ *
+ *  If +count+ is zero, returns +self+ unmodified:
+ *    a = [:foo, 'bar', 2]
+ *    a.rotate!(0)
+ *    a # => [:foo, "bar", 2]
+ *
+ *  ---
+ *
+ *  When given a negative +count+, rotates in the opposite direction,
+ *  from end to beginning:
+ *    a = [:foo, 'bar', 2]
+ *    a.rotate!(-2)
+ *    a # => ["bar", 2, :foo]
+ *
+ *  If +count+ is small, uses <tt>count % ary.size</tt> as the count:
```

Comment:
My first thought seeing "small" here made me think "close to zero", so the modulo arithmetic confused me

I'm not sure if something like "is a large negative number" or combining the positive and negative cases in one explanation would clear that up

Maybe it's not worth fixing?

Reply:
Well, I thought the same thing as I wrote it.  But (former mathematics  teacher that I am), I not e that the largest negative integer is -1.  How a bout "small" => "far from zero"?   We're already in a marked out section on negative +count+, so would that make sense?

Reply:
Yeah, I know it's mathematically correct, but I tripped over myself.  If you decide to make a change I like "far from zero".

Reply:
Added "(far from zero)".